resin/logs.py

```diff
@@ -1,133 +1,189 @@
+from twisted.web.client import Agent
+from twisted.web.http_headers import Headers
+from twisted.internet.ssl import ClientContextFactory
+
+from threading import Thread
 
 from .base_request import BaseRequest
+from . import exceptions
 from .models.config import Config
 from .models.device import Device
+from .settings import Settings
+
+
+class WebClientContextFactory(ClientContextFactory):
+    """
+    This is low level class and is not meant to be used by end users directly.
+    """
 
+    def getContext(self, hostname, port):
+        return ClientContextFactory.getContext(self)
 
-logging.basicConfig(format='%(levelname)s:%(message)s')
 
+class StreamingParser(Protocol):
+    """
+    This is low level class and is not meant to be used by end users directly.
+    """
 
-def _print_deprecation_warning():
-    logging.warning(" using legacy logging services, this will stop working shortly.\nPlease update to ensure logs are correctly retrieved in future.")
+    def __init__(self, callback, error):
+        self.callback = callback
+        self.error = error
 
+    def dataReceived(self, data):
+        d = Deferred()
+        d.addCallback(self.display_data)
+        d.addErrback(self.error_func)
+        d.callback(data)
 
-# TODO: https://github.com/resin-io/resin-sdk/pull/277/files
-class Logs(object):
+    def connectionLost(self, reason):
+        pass
+
+    def display_data(self, data):
+        lines = data.splitlines()
+        for line in lines:
+            try:
+                self.callback(json.loads(line))
+            except:
+                # Ignore incomplete log entry, this can happen with the old logs
+                pass
+
+    def error_func(self, error):
+        if self.error:
+            d = Deferred()
+            d.addErrback(self.error)
+            d.errback(DefaultException(error))
+
+
+def cbRequest(response, callback, error):
+    protocol = StreamingParser(callback, error)
+    response.deliverBody(protocol)
+    return protocol
+
+
+def cbDrop(protocol):
+    protocol.transport.stopProducing()
+    protocol.transport.loseConnection()
+
+
+class Subscription(object):
     """
-    This class implements functions that allow processing logs from device.
+    This is low level class and is not meant to be used by end users directly.
+    """
+
+    def __init__(self):
+        self.context_factory = WebClientContextFactory()
+        self.settings = Settings()
+        super(Subscription, self).__init__()
+
+    def add(self, uuid, callback, error=None, count=None):
+
+        query = 'stream=1'
+        if count:
+            query = 'stream=1&count={}'.format(count)
 
-    This class is implemented using pubnub python sdk.
+        url = urljoin(
+            self.settings.get('api_endpoint'),
+            '/device/v2/{uuid}/logs?{query}'.format(uuid=uuid, query=query)
+        )
+        headers = {}
+        headers[b'Authorization'] = ['Bearer {:s}'.format(self.settings.get('token')).encode()]
```

Comment:
Why is the value of the header a list?

Reply:
it's how twisted defines a header, `_rawHeaders | A dict mapping header names as str to lists of header values as str.` (https://twistedmatrix.com/documents/9.0.0/api/twisted.web.http_headers.Headers.html)


`